File: sample-apps/objc-sample-app/objc-sample-app/DeeplinkHandler.m
//
//  DeeplinkHandler.m
//  objc-sample-app
//
//  Created by Tapash Majumder on 6/21/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

#import "DeeplinkHandler.h"
#import "CoffeeType.h"
#import "CoffeeViewController.h"
#import "CoffeeListTableViewController.h"

@implementation DeeplinkHandler

+ (BOOL)handleURL:(NSURL *)url {
    NSString *page = url.lastPathComponent.lowercaseString;
    
    if ([page isEqualToString:@"mocha"]) {
        [DeeplinkHandler showCoffee:CoffeeType.mocha];
        return YES;
    } else if ([page isEqualToString:@"latte"]) {
        [DeeplinkHandler showCoffee:CoffeeType.latte];
        return YES;
    } else if ([page isEqualToString:@"cappuccino"]) {
        [DeeplinkHandler showCoffee:CoffeeType.cappuccino];
        return YES;
    } else if ([page isEqualToString:@"black"]) {
        [DeeplinkHandler showCoffee:CoffeeType.black];
        return YES;
    } else if ([page isEqualToString:@"coffee"]) {
        NSString *query = [DeeplinkHandler parseQueryFromURL: url];
        [DeeplinkHandler showCoffeeListWithQuery: query];
        return YES;
    } else {
        [UIApplication.sharedApplication openURL: url options: @{} completionHandler: nil];
        return NO;
    }
}

+ (void)showCoffee:(CoffeeType *)coffeeType {
    UINavigationController *rootNav = (UINavigationController *) UIApplication.sharedApplication.keyWindow.rootViewController;
    if (rootNav != nil) {
        [rootNav popToRootViewControllerAnimated: false];

        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle: nil];
        CoffeeViewController *viewController = (CoffeeViewController *) [storyboard instantiateViewControllerWithIdentifier: @"CoffeeViewController"];
        viewController.coffeeType = coffeeType;
        
        [rootNav pushViewController: viewController animated: true];
    }
}

+ (void)showCoffeeListWithQuery:(NSString *)query {
    UINavigationController *rootNav = (UINavigationController *) UIApplication.sharedApplication.keyWindow.rootViewController;

    if (rootNav != nil) {
        [rootNav popToRootViewControllerAnimated: true];
        CoffeeListTableViewController *coffeeListVC = (CoffeeListTableViewController *) rootNav.viewControllers[0];
        if (coffeeListVC != nil) {
            coffeeListVC.searchTerm = query;
        }
    }
}

+ (NSString *)parseQueryFromURL:(NSURL *)url {
    NSURLComponents *components = [[NSURLComponents alloc] initWithURL: url resolvingAgainstBaseURL: false];
    
    if (components == nil || components.queryItems == nil) {
        return nil;
    }
    
    NSUInteger index = [components.queryItems indexOfObjectPassingTest:^BOOL(NSURLQueryItem * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        return [obj.name isEqualToString: @"q"];
    }];
    
    if (index == NSNotFound) {
        return nil;
    } else {
        return components.queryItems[index].value;
    }
}
@end

File: sample-apps/swift-sample-app/swift-sample-app/DeeplinkHandler.swift
//
//  DeeplinkHandler.swift
//  iOS Demo
//
//  Created by Tapash Majumder on 5/18/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

import Foundation
import UIKit

import IterableSDK

struct DeeplinkHandler {
    static func handle(url: URL) -> Bool {
        if let deeplink = Deeplink.from(url: url) {
            show(deeplink: deeplink)
            return true
        } else {
            return false
        }
    }
    
    private static func show(deeplink: Deeplink) {
        if let coffeeType = deeplink.toCoffeeType() {
            // single coffee
            show(coffee: coffeeType)
        } else {
            // coffee list with query
            if case let .coffee(query) = deeplink {
                showCoffeeList(query: query)
            } else {
                assertionFailure("could not determine coffee type.")
            }
        }
    }
    
    private static func show(coffee: CoffeeType) {
        let coffeeVC = CoffeeViewController.createFromStoryboard()
        coffeeVC.coffee = coffee
        if let rootNav = UIApplication.shared.delegate?.window??.rootViewController as? UINavigationController {
            if let coffeeListVC = rootNav.viewControllers[0] as? CoffeeListTableViewController {
                coffeeListVC.searchTerm = nil
            }
            rootNav.popToRootViewController(animated: false)
            rootNav.pushViewController(coffeeVC, animated: true)
        }
    }
    
    private static func showCoffeeList(query: String?) {
        if let rootNav = UIApplication.shared.delegate?.window??.rootViewController as? UINavigationController {
            rootNav.popToRootViewController(animated: true)
            if let coffeeListVC = rootNav.viewControllers[0] as? CoffeeListTableViewController {
                coffeeListVC.searchTerm = query
            }
        }
    }
    
    // This enum helps with parsing of Deeplinks.
    // Given a URL this enum will return a Deeplink.
    // The deep link comes in as http://domain.com/../mocha
    // or http://domain.com/../coffee?q=mo
    private enum Deeplink {
        case mocha
        case latte
        case cappuccino
        case black
        case coffee(q: String?)
        
        static func from(url: URL) -> Deeplink? {
            let page = url.lastPathComponent.lowercased()
            switch page {
            case "mocha":
                return .mocha
            case "latte":
                return .latte
            case "cappuccino":
                return .cappuccino
            case "black":
                return .black
            case "coffee":
                return parseCoffeeList(fromUrl: url)
            default:
                return nil
            }
        }
        
        private static func parseCoffeeList(fromUrl url: URL) -> Deeplink {
            guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
                return .coffee(q: nil)
            }
            guard let queryItems = components.queryItems else {
                return .coffee(q: nil)
            }
            guard let index = queryItems.firstIndex(where: { $0.name == "q" }) else {
                return .coffee(q: nil)
            }
            
            return .coffee(q: queryItems[index].value)
        }
        
        // converts deep link to coffee
        // return nil if it refers to coffee list
        func toCoffeeType() -> CoffeeType? {
            switch self {
            case .coffee:
                return nil
            case .black:
                return .black
            case .cappuccino:
                return .cappuccino
            case .latte:
                return .latte
            case .mocha:
                return .mocha
            }
        }
    }
}
